executor/runcexecutor/executor_linux.go

```diff
@@ -0,0 +1,308 @@
+	"github.com/containerd/containerd/sys/reaper"
+	runc "github.com/containerd/go-runc"
+	"github.com/moby/buildkit/executor"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sys/unix"
+)
+
+var setupReaper sync.Once
+
+func handleSignals(ctx context.Context) error {
+	signals := make(chan os.Signal, 32)
+	signal.Notify(signals, unix.SIGPIPE, unix.SIGCHLD)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case s := <-signals:
+			switch s {
+			case unix.SIGCHLD:
+				if err := reaper.Reap(); err != nil {
+					logrus.WithError(err).Error("reap exit status")
+				}
+			case unix.SIGPIPE:
+			}
+		}
+	}
+}
+
+func (w *runcExecutor) initPlatform() (*runcExecutor, error) {
+	setupReaper.Do(func() {
+		runc.Monitor = reaper.Default
+		// collect sigchld from grandchildren processes so we can see when
+		// the forked/detached runc processes exit
+		reaper.SetSubreaper(1)
+		go handleSignals(context.Background())
+	})
+
+	stdioPlatform, err := runcV2.NewPlatform()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to initialize stdio platform")
+	}
+	w.stdioPlatform = stdioPlatform
+	return w, nil
+}
+
+func (w *runcExecutor) run(ctx context.Context, id, bundle string, process executor.ProcessInfo, ended chan struct{}) (int, error) {
+	var endedOnce sync.Once
+	defer endedOnce.Do(func() {
+		close(ended)
+	})
+	pidfile, err := ioutil.TempFile(bundle, "run-*.pid")
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "failed to create pidfile")
+	}
+	pidfile.Close()
+
+	createOpts := &runc.CreateOpts{
+		NoPivot: w.noPivot,
+		PidFile: pidfile.Name(),
+		Detach:  true,
+	}
+
+	initErrCh := make(chan error)
+	if process.Meta.Tty {
+		var socket runc.ConsoleSocket
+		var cleanup func()
+		socket, cleanup, err = w.setupConsole(ctx, initErrCh, ended, bundle, id, process)
+		if err != nil {
+			return containerd.UnknownExitStatus, err
+		}
+		defer cleanup()
+		createOpts.ConsoleSocket = socket
+	} else {
+		close(initErrCh)
+		createOpts.IO = &forwardIO{stdin: process.Stdin, stdout: process.Stdout, stderr: process.Stderr}
+	}
+
+	exitCh := reaper.Default.Subscribe()
+
+	eg, runCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		_, err = w.runc.Run(runCtx, id, bundle, createOpts)
+		return err
+	})
+	eg.Go(func() error {
+		return <-initErrCh
+	})
+	err = eg.Wait()
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "runc create failed")
+	}
+
+	pid, err := runc.ReadPidFile(pidfile.Name())
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "failed to read runc pidfile")
+	}
+
+	status := waitpid(ctx, pid, exitCh)
+	defer endedOnce.Do(func() {
+		close(ended)
+	})
+	return status, nil
+}
+
+func (w *runcExecutor) exec(ctx context.Context, id, bundle string, pid0 int, specsProcess *specs.Process, process executor.ProcessInfo) (int, error) {
```

Comment:
A lot of this seems duplicated with `(*runcExecutor).run`.

Reply:
Yeah, I will take a look to try to dedup.